source.go

```diff
@@ -53,46 +70,63 @@ func (s *Source) Configure(ctx context.Context, cfg map[string]string) error {
 }
 
 func (s *Source) Open(ctx context.Context, pos sdk.Position) error {
-	// Open is called after Configure to signal the plugin it can prepare to
-	// start producing records. If needed, the plugin should open connections in
-	// this function. The position parameter will contain the position of the
-	// last record that was successfully processed, Source should therefore
-	// start producing records after this position. The context passed to Open
-	// will be cancelled once the plugin receives a stop signal from Conduit.
+	sdk.Logger(ctx).Info().Msg("Opening Source...")
+	s.server = source.NewServer(ctx)
+	err := s.runServer(ctx)
+	if err != nil {
+		return err
+	}
 	return nil
 }
 
 func (s *Source) Read(ctx context.Context) (sdk.Record, error) {
-	// Read returns a new Record and is supposed to block until there is either
-	// a new record or the context gets cancelled. It can also return the error
-	// ErrBackoffRetry to signal to the SDK it should call Read again with a
-	// backoff retry.
-	// If Read receives a cancelled context or the context is cancelled while
-	// Read is running it must stop retrieving new records from the source
-	// system and start returning records that have already been buffered. If
-	// there are no buffered records left Read must return the context error to
-	// signal a graceful stop. If Read returns ErrBackoffRetry while the context
-	// is cancelled it will also signal that there are no records left and Read
-	// won't be called again.
-	// After Read returns an error the function won't be called again (except if
-	// the error is ErrBackoffRetry, as mentioned above).
-	// Read can be called concurrently with Ack.
-	return sdk.Record{}, nil
+	select {
+	case <-ctx.Done():
+		return sdk.Record{}, ctx.Err()
+	case record, ok := <-s.server.RecordCh:
+		if !ok {
+			return sdk.Record{}, fmt.Errorf("record channel is closed")
+		}
+		return record, nil
+	}
 }
 
 func (s *Source) Ack(ctx context.Context, position sdk.Position) error {
-	// Ack signals to the implementation that the record with the supplied
-	// position was successfully processed. This method might be called after
-	// the context of Read is already cancelled, since there might be
-	// outstanding acks that need to be delivered. When Teardown is called it is
-	// guaranteed there won't be any more calls to Ack.
-	// Ack can be called concurrently with Read.
-	return nil
+	sdk.Logger(ctx).Debug().Str("position", string(position)).Msg("got ack")
+	return s.server.SendAck(position)
 }
 
 func (s *Source) Teardown(ctx context.Context) error {
-	// Teardown signals to the plugin that there will be no more calls to any
-	// other function. After Teardown returns, the plugin should be ready for a
-	// graceful shutdown.
+	if s.server != nil {
+		// server might be waiting for this signal, if pipeline was stopped.
+		s.server.Teardown <- true
```

Comment:
Nitpick: it's common to close the channel rather than send a value into it to signal a close. It works pretty much the same though. Well, the one difference is that multiple goroutines can see that a channel is closed, while the value sent to the channel would only be received by one goroutine. We only have one goroutine now, so it doesn't make a difference, but if we plan to add support for multiple streams in the future we might as well change this now to prevent issues.

Reply:
interesting! thanks for the suggestion